tensorflow/java/src/main/native/session_jni.cc

```diff
@@ -113,26 +113,43 @@ JNIEXPORT jlong JNICALL Java_org_tensorflow_Session_allocate2(
     throwException(env, kNullPointerException, "Graph has been close()d");
     return 0;
   }
-  TF_Graph* graph = reinterpret_cast<TF_Graph*>(graph_handle);
+
```

Comment:
Why the changes to this file?

Reply:
Right, but the change here seems more than just plugging those leaks. Why bother with `TF_NewBufferFromString`? Can't we reduce the change in this file to 2 lines:

```c++
if (!throwExcpetionIfNotOK(env, status)) {
  env->ReleaseByteArrayElemenst(config, cconfig, JNI_ABORT);
  TF_DeleteSessionOptions(opts);
  TF_DeleteStatus(status);
  return 0;
}
```
?

Reply:
Yes, it look simple but I have some reasons as blow.
1.  Keep the code same, please take a look for saved_model_bundle_jni.cc code for run_options logic
`  if (run_options != nullptr) {
    size_t sz = env->GetArrayLength(run_options);
    if (sz > 0) {
      jbyte* run_options_data = env->GetByteArrayElements(run_options, nullptr);
      crun_options =
          TF_NewBufferFromString(static_cast<void*>(run_options_data), sz);
      env->ReleaseByteArrayElements(run_options, run_options_data, JNI_ABORT);
    }
  }`
2. I concern for method TF_SetConfig `options.config.ParseFromArray(proto, proto_len)`. It is protocol buffer logic,  I am not sure it use or copy the proto bytearray into options.config. The last logic will delete options, the proto bytearray should be released in java garbage collect.

Reply:
Thanks, but I'm not sure I follow.

1. `saved_model_bundle_jni.cc` has to use `TF_NewBufferFromString` because the C API function it eventually invokes (`TF_LoadSessionFromSavedModel`) requires a `TF_Buffer*` argument. That is not the case here.

2. `TF_SetConfig` does not hold on to memory of the passed in arguments upon returning, so I don't think you need to be concerned.